### ISSUE TITLE
How to show errors in Builder when GeoCoding limits are exceeded

### ISSUE THREAD

**Issue by jorgesancha** [CartoDB/dataservices-api#368]
Currently it is possible to set limits to the number of Geocodings a user, organization or a given server can perform in a given period of time. See here for more details:

https://github.com/CartoDB/dataservices-api/blob/master/doc/rate_limits.md

As @jgoizueta mentions in [this comment](https://github.com/CartoDB/dataservices-api/issues/346#issuecomment-290763912), if the limits are exceeded when performing Geocoding in Builder, the API fails silently and the analysis continues returning null values for those rows that could not be geocoded. This is to ensure you make the best of the remaining geocodings you have: if you are trying to geocode 5000 rows and you only have 3000 geocodings left, it will geocode 3000 and return null for the remaining 2000.

However, when using the API directly, a Rate-limit-exceeded error is returned.

I'd like to understand what would be required in order to show an error or warning in Builder if the limits are exceeded, whether it is currently possible as the functionality stands or whether further development would be required.

**Comment by jorgesancha** [comment]
cc @noguerol

**Comment by jorgesancha** [comment]
By the way, I'm assuming here that this does not use the same Quotas/Credits mechanism we use for Data Services in general, but I might be mistaken?

Or to put it differently, if I were to run my 5000 row analysis, would I get today a warning that I don't have enough geocodings to run it?

**Comment by rafatower** [comment]
> By the way, I'm assuming here that this does not use the same Quotas/Credits mechanism we use for Data Services in general, but I might be mistaken?

You're right. Quota and rate limits are different things with different product and technical requirements.

> Or to put it differently, if I were to run my 5000 row analysis, would I get today a warning that I don't have enough geocodings to run it?

Short answer: no, you would not get anything from the UI to tell that you "exceeded your rate limit".

Longer answer:

By design, when the quota or rate limits are exceeded the DS API will throw an `ERROR` at DB level.

The effect of this when executed from within any query is that the whole transaction is rolled back.

In order to workaround that behavior and better integrate it into Builder (actually within Engine, and more specifically camshaft, that makes usage of the SQL Batch API) we implemented what we called [exception-safe functions](https://github.com/CartoDB/dataservices-api/blob/development/doc/internal/exception_safe.md). What they do is basically wrap around the vanilla functions and convert any exception thrown from within the original function into a `WARNING`.

Let me clarify the call flows and interactions with a diagram:

![image](https://cloud.githubusercontent.com/assets/1914276/26352794/a0e84d02-3fbd-11e7-9e77-b1501ee53dc2.png)

However, if you use the exception-safe flavor of the functions, this is the behavior:

![image](https://cloud.githubusercontent.com/assets/1914276/26352884/0f6711dc-3fbe-11e7-8243-c06a231e6790.png)

(I ignore whether the Batch SQL API is able to capture those warnings, but the plain SQL API does so I assume there's no limitation)


All this imposes some inherent technical limitations:
- We cannot know about the scope of the original transaction from within the DS API Server.
- Any `ERROR` in DB (a.k.a. exception), by default, will always rollback the current transaction. Typical transactions launched from the analysis layer (as it stands to date) are the creation of new datasets. They turn out to be all-or-nothing operations, from a DB point of view.
- Any `ERROR` can only be dealt with from within the DB by using some PL language (pl/pgsql, pl/python or else).

To sum up, in order to show proper feedback in the UI (for rate limits and for quotas), we'd need to propagate the information conveyed in `WARNING`'s from DB back to the user throughout all layers depicted above, possibly with a bit of processing to be done by the analysis layer (in your example there would be 2000 warnings).

**Comment by rafatower** [comment]
Closing this FTM, feel free to reopen if there are further questions.
